### PR TITLE
New Recipe: Location based alerting GH-114

### LINKED ISSUE #114: Location Based Alerting Demo (geohash version)
**Describe the use case**
https://github.com/confluentinc/apac-workshops/tree/d134877b5587bdfc842c7231abbb96fa92e30420/marketing-locationbased-geohash

Show real time, personalized location based alerts in real time. User and merchant data from a database is joined against real time user locations (e.g. from mobile devices) to provide alerts when users pass within 200 meters of a participating merchant.  User locations are subdivided by geohash and users receive an alert if they come within 200m of any participating merchant (the red circle).

**Provide the ksqlDB application**
Document the full set of ksqlDB statements that represent the use case.

https://github.com/confluentinc/apac-workshops/blob/master/marketing-locationbased-geohash/ksql-statements.sql

### PR DESCRIPTION
**Pull request by rspurgeon**
### Description 
Resolves #114 

### New recipe checklist
- [X] Full code validated in Confluent Cloud
- [X] SQL style/syntax consistent to existing recipes
- [X] Validate presentation with `python3 -m mkdocs serve`
- [X] File structure and content follow [template](https://github.com/confluentinc/ksqldb-recipes/tree/main/template)
- [X] Recipe added to [navigation](https://github.com/confluentinc/ksqldb-recipes/blob/main/mkdocs.yml): comment out until phase release
- [X] Spell check, e.g. using `aspell`
- [X] Source connector's auto topic naming convention works with the ksqlDB app values for `KAFKA_TOPIC`

### CODE REVIEW

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/process.sql:46]: This topic `alerts-raw` is empty (`user-locations` is not)

**Review by ybyzek — CHANGES_REQUESTED** [review]
Thanks @rspurgeon !  Not sure what's happening but one of the queries isn't working -- empty topic (please see comments)

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql:45]: @ybyzek on my test I see this:
```
PRINT 'alerts-raw' FROM BEGINNING;
Key format: ¯\_(ツ)_/¯ - no data processed
Value format: JSON or KAFKA_STRING
rowtime: 2022/01/26 14:40:24.691 Z, key: <null>, value: {"USER_ID":2,"USER_LATITUDE":14.5517401,"USER_LONGITUDE":121.0518652,"USER_GEOHASH":"wdw4f9","MERCHANT_ID":3}, partition: 4
rowtime: 2022/01/26 14:40:24.806 Z, key: <null>, value: {"USER_ID":2,"USER_LATITUDE":14.5500341,"USER_LONGITUDE":121.0555802,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":2}, partition: 4
rowtime: 2022/01/26 14:40:24.806 Z, key: <null>, value: {"USER_ID":2,"USER_LATITUDE":14.5500341,"USER_LONGITUDE":121.0555802,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":1}, partition: 4
rowtime: 2022/01/26 14:40:24.494 Z, key: <null>, value: {"USER_ID":1,"USER_LATITUDE":14.5472791,"USER_LONGITUDE":121.0475401,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":2}, partition: 3
rowtime: 2022/01/26 14:40:24.494 Z, key: <null>, value: {"USER_ID":1,"USER_LATITUDE":14.5472791,"USER_LONGITUDE":121.0475401,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":1}, partition: 3
rowtime: 2022/01/26 14:40:24.602 Z, key: <null>, value: {"USER_ID":2,"USER_LATITUDE":14.5486952,"USER_LONGITUDE":121.0521851,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":2}, partition: 3
rowtime: 2022/01/26 14:40:24.602 Z, key: <null>, value: {"USER_ID":2,"USER_LATITUDE":14.5486952,"USER_LONGITUDE":121.0521851,"USER_GEOHASH":"wdw4f8","MERCHANT_ID":1}, partition: 3
```

Comment (ybyzek) [reply docs/real-time-analytics/location-based-alerting/process.sql:45]: Now resolved

File: docs/real-time-analytics/location-based-alerting/index.md
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/index.md:33]: nit `into a ksqlDB [t]able keyed by`

Comment (mjsax) [reply docs/real-time-analytics/location-based-alerting/index.md:33]: Geohash or geohash (why capital G ?)

Comment (mjsax) [reply docs/real-time-analytics/location-based-alerting/index.md:33]: nit `As the user[']s location changes`

Comment (mjsax) [reply docs/real-time-analytics/location-based-alerting/index.md:33]: > and matches result in a stream

Don't understand this part? It is grammatically correct (not a native speaker)?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:16]: > on the a substring

`the` or `a` ?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:58]: Why do we need this clause?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:75]: nit: formatting of the above 4 lines (hard to read)

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:76]: `"lat"` -- Why use quotes? Just `lat` should also work. (same for `"lon"`)

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:81]: nit: as above -- formatting -- hard to read

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [inline docs/real-time-analytics/location-based-alerting/process.sql:82]: as above: why do we need this?

File: docs/real-time-analytics/location-based-alerting/index.md
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/index.md:33]: > Geohash or geohash (why capital G ?)

Followed a convention found here: https://en.wikipedia.org/wiki/Geohash

Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/index.md]: @mjsax please see revised copy.  Thanks for the review

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql]: Thank you, resolved

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql:58]: I wasn't sure either, the examples were based off the associated demo in the GH issue, here: https://github.com/confluentinc/ksqldb-recipes/issues/114

Should this query work without specifying the `PARTITION BY` ?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql]: @mjsax please see revised code, hopefully that's easier to read

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql]: Also, based on the associated demo. Removed in latest

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql:82]: As above: based on the associated GH issue demo. Is this required for the query to work?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql]: Please see latest and LMK if that's easier to read or no help... thanks

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [reply docs/real-time-analytics/location-based-alerting/process.sql:58]: I believe yes (but there could also be something about ksqlDB that I don't know of...) -- Mind trying it out and let me know?

**Review by fifthposition — COMMENTED** [review]
Copyedited.

**Review by ybyzek — APPROVED** [review]
@rspurgeon thanks for the recipe, LGTM, if you could please also address the last open comment from @mjsax https://github.com/confluentinc/ksqldb-recipes/pull/159/files#r793197772 before merge

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (ybyzek) [reply docs/real-time-analytics/location-based-alerting/process.sql:58]: @rspurgeon can you review this before merge?

**Review by ybyzek — CHANGES_REQUESTED** [review]
@rspurgeon I ran through this sequence again (making sure that `earliest` was set) and again hit https://github.com/confluentinc/ksqldb-recipes/pull/159#discussion_r792793689 .  It wasn't until I added the manual `INSERT INTO` twice that there was output.  Can you please take a look and see if you can repro this on your end?  (cc: @mjsax for ideas)

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql:58]: Removing `PARTITION BY null` results in:
```
Could not determine output schema for query due to error: Key missing from projection. The query used to build `ALERTS_RAW` must include the join expression MERCHANTS_BY_GEOHASH.GEOHASH in its projection. Statement: CREATE STREAM ALERTS_RAW WITH (KAFKA_TOPIC='alerts-raw', PARTITIONS=6, REPLICAS=3, VALUE_FORMAT='JSON') AS SELECT USER_LOCATIONS.ID USER_ID, USER_LOCATIONS.LATITUDE USER_LATITUDE, USER_LOCATIONS.LONGITUDE USER_LONGITUDE, SUBSTRING(USER_LOCATIONS.GEOHASH, 1, 6) USER_GEOHASH, EXPLODE(MERCHANTS_BY_GEOHASH.ID_LIST) MERCHANT_ID FROM USER_LOCATIONS USER_LOCATIONS LEFT OUTER JOIN MERCHANTS_BY_GEOHASH MERCHANTS_BY_GEOHASH ON ((SUBSTRING(USER_LOCATIONS.GEOHASH, 1, 6) = MERCHANTS_BY_GEOHASH.GEOHASH)) EMIT CHANGES;
```

**Comment by rspurgeon** [comment]
> @rspurgeon I ran through this sequence again (making sure that `earliest` was set) and again hit [#159 (comment)](https://github.com/confluentinc/ksqldb-recipes/pull/159#discussion_r792793689) . It wasn't until I added the set of manual `INSERT INTO` **twice** that there was output. Can you please take a look and see if you can repro this on your end? (cc: @mjsax for ideas)

@ybyzek I can't reproduce. I started an entirely new env, cluster, and ksql app and when I insert the 5 records in the manual file and I get output data in the `alerts_raw` stream

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (mjsax) [reply docs/real-time-analytics/location-based-alerting/process.sql:58]: Thanks. I understand.

It's something we need to fix on our side... It's an unnecessary required that a result STREAM needs a key.

**Comment by ybyzek** [comment]
https://github.com/confluentinc/ksqldb-recipes/pull/159#issuecomment-1027374311

@rspurgeon I just reproduced this a third time!  Want to jump on a zoom?

**Comment by ybyzek** [comment]
Update: we both reproduced the issue on a call, posting question to ksqlDB team in [Slack](https://confluent.slack.com/archives/C337JP2F8/p1643821068712449)

File: docs/real-time-analytics/location-based-alerting/manual.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/manual.sql:9]:
```suggestion
-- Wait 10 seconds before inserting the records below records
```

File: docs/real-time-analytics/location-based-alerting/index.md
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/index.md:40]: @rspurgeon Add another note here?

File: docs/real-time-analytics/location-based-alerting/manual.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/manual.sql:9]:
```suggestion
-- Wait 10 seconds before inserting the records below
```

Sorry!

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/process.sql:49]:
```suggestion
SET 'max.task.idle.ms' = '30000';
```

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/process.sql:46]:
```suggestion
--    we ensure that the merchants_by_geohash table receives data prior to the 
```

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (ybyzek) [inline docs/real-time-analytics/location-based-alerting/process.sql:42]: @mjsax can you please review & validate this explanation?

File: docs/real-time-analytics/location-based-alerting/process.sql
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/process.sql]: resolved already

File: docs/real-time-analytics/location-based-alerting/index.md
Comment (rspurgeon) [reply docs/real-time-analytics/location-based-alerting/index.md:40]: added in latest

**Review by ybyzek — APPROVED** [review]
LGTM thanks @rspurgeon

**Comment by rspurgeon** [comment]
rebased off `main` and force pushed

**Pull request merged**
